### PR TITLE
COSE receipts: CPP API

### LINKED ISSUE #6477: Support COSE receipts in CCF
```[tasklist]
### Tasks
- [ ] get_cose_sign1_signature() API in historical queries
- [ ] get_cbor_merkle_proof() API in historical queries
```

### PR DESCRIPTION
**Pull request by maxtropets**
Closes #6477

### CODE REVIEW

File: src/node/snapshot_serdes.h [hunk 1]
```diff
@@ -161,6 +161,7 @@ namespace ccf
     cd.set(std::move(claims_digest));
     ccf::TxReceiptImpl tx_receipt(
       sig,
+      std::nullopt, // cose
```
Comment (maxtropets) [inline src/node/snapshot_serdes.h:164]: Shall we have in a snapshot serdes at the moment?

Comment (achamayou) [reply src/node/snapshot_serdes.h:164]: Yes, let's be as uniform as possible. Ideally this is step 1 of a 3 step plan:

1. Add COSE signatures as something the application code can use
2. Rebase all internal and tool usage to COSE signatures
3. Offer a configuration option to start new ledgers "COSE only"

File: src/node/historical_queries_adapter.cpp [hunk 2]
```diff
@@ -153,6 +162,37 @@ namespace ccf
...
+  {
+    nlohmann::json out = nlohmann::json();
+    static constexpr int LEAF_LABEL = 404; // TBD
+    static constexpr int PROOF_LABEL = 404 + 1; // TBD
+    static constexpr bool TREE_PATH_LEFT = 0;
+    static constexpr bool TREE_PATH_RIGHT = 1;
+
+    out[LEAF_LABEL] = {
+      nlohmann::json::binary_t(from_string(in.write_set_digest->hex_str())),
+      nlohmann::json::binary_t(from_string(in.commit_evidence.value())),
```
Comment (achamayou) [inline src/node/historical_queries_adapter.cpp:176]: This needs to be `tstr`

Comment (achamayou) [inline src/node/historical_queries_adapter.cpp:176]:
```suggestion
      nlohmann::json::binary_t(in.commit_evidence.value()),
```

File: src/node/historical_queries_adapter.cpp [hunk 3]
```diff
@@ -153,6 +162,37 @@ namespace ccf
...
+      nlohmann::json::binary_t(
+        from_string(in.claims_digest.value().hex_str()))};
+
+    auto path = nlohmann::json::array();
+    for (const auto& node : *in.path)
+    {
+      auto n = nlohmann::json::object();
+      if (node.direction == ccf::HistoryTree::Path::Direction::PATH_LEFT)
+      {
+        path[TREE_PATH_LEFT] = node.hash.to_string();
```
Comment (achamayou) [inline src/node/historical_queries_adapter.cpp:186]: That's probably going to be tstr?

File: src/node/snapshot_serdes.h (hunk 1 above)
Comment (maxtropets) [reply src/node/snapshot_serdes.h:164]: But this means we have to add a cose signature to snapshot, doesn't it? Do we do it in this PR then?

Comment (maxtropets) [reply src/node/snapshot_serdes.h:164]: Talked about that, discussed to leave as is for now, will do in subsequent PRs.

File: src/node/historical_queries_adapter.cpp [hunk 4]
```diff
@@ -10,6 +10,58 @@
 #include "node/rpc/network_identity_subsystem.h"
 #include "node/tx_receipt_impl.h"
 
+#include <t_cose/t_cose_sign1_sign.h>
+
+namespace
+{
+  static constexpr int64_t LEAF_LABEL = 404; // TBD
```
Comment (achamayou) [inline src/node/historical_queries_adapter.cpp:17]: Adding the link to https://github.com/ietf-scitt/draft-birkholz-cose-cometre-ccf-profile in comment is probably good.

File: src/node/historical_queries_adapter.cpp [hunk 5]
```diff
@@ -10,6 +10,58 @@
...
+  static constexpr int64_t PATH_LABEL = 404 + 1; // TBD
+
+  std::vector<uint8_t> from_string(std::string_view s)
```
Comment (achamayou) [inline src/node/historical_queries_adapter.cpp:20]: suggest we move that to ds, it's often useful

File: src/node/historical_queries_adapter.cpp [hunk 6]
```diff
@@ -153,6 +205,49 @@ namespace ccf
...
+
+    struct q_useful_buf_c result;
+    auto qerr = QCBOREncode_Finish(&ctx, &result);
+    if (qerr)
+    {
+      throw std::logic_error(
+        fmt::format("Failed to encode merkle proof: {}", qerr));
+    }
+
+    return {
```
Comment (maxtropets) [inline src/node/historical_queries_adapter.cpp:247]: Check out the possibility to re-use buffer memory

**Review by achamayou — APPROVED** [review]

File: src/node/historical_queries_adapter.cpp [hunk 7]
```diff
@@ -10,6 +10,58 @@
...
+    return {data, data + s.size()};
+  }
+
+  void encode_leaf_cbor(
+    QCBOREncodeContext& ctx, const ccf::TxReceiptImpl& receipt)
+  {
+    QCBOREncode_OpenArrayInMapN(&ctx, LEAF_LABEL);
+
+    // 1 WSD
+    const auto wsd = from_string(receipt.write_set_digest->hex_str());
```
Comment (maxtropets) [inline src/node/historical_queries_adapter.cpp:32]: WTF change

File: src/node/historical_queries_adapter.cpp [hunk 8]
```diff
@@ -10,6 +10,58 @@
...
+    QCBOREncode_AddBytes(&ctx, {wsd.data(), wsd.size()});
+
+    // 2. CE
+    const auto ce = receipt.commit_evidence.value();
+    QCBOREncode_AddSZString(&ctx, ce.data());
+
+    // 3. CD
+    const auto cd = from_string(receipt.claims_digest.value().hex_str());
```
Comment (maxtropets) [inline src/node/historical_queries_adapter.cpp:40]: Change

File: src/node/historical_queries_adapter.cpp (hunk 5 above)
Comment (maxtropets) [reply src/node/historical_queries_adapter.cpp:17]: It's actually hasn't been used in this PR anymore, so I deleted it..

File: src/node/historical_queries_adapter.cpp (hunk 6 above)
Comment (maxtropets) [reply src/node/historical_queries_adapter.cpp]: They're actually same, both for `cose_sign1` and here for encode/decode.

But now I think, for `cose_sign1()` you'll have to move buffer size estimation outside, and same for verification, if you want to create buffer in place and pass it inside.

@achamayou  I think it's OK to leave it as is and only copy-paste the decode/encode part, looks simpler.

**Comment by maxtropets** [comment]
Proof example

HEX: `8dfa2190194835820f70ab02a71151d04ed5009907cc3b2cb2c7e9a7a8fb3f66a7ee73cd5b7ed2bf8784863653a322e31323a37373533663833306364626166323562303739303265326230666464643863316665313936623034613836323536323961333963373338656263376237343836582000000000000000000000000000000000000000000000000000000000000000001901958282015820f45e9165fafda60a24550e76df74b75bec70a83e8d6437e2d7c23c5949242e27820158204f5407075ce5df9b6208956ff15723e16b39591e6e763f5bc257363871e4074f`

bytes.fromhex(): `b"X\xdf\xa2\x19\x01\x94\x83X \xf7\n\xb0*q\x15\x1d\x04\xedP\t\x90|\xc3\xb2\xcb,~\x9az\x8f\xb3\xf6j~\xe7<\xd5\xb7\xed+\xf8xHce:2.12:7753f830cdbaf25b07902e2b0fddd8c1fe196b04a8625629a39c738ebc7b7486X \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x19\x01\x95\x82\x82\x01X \xf4^\x91e\xfa\xfd\xa6\n$U\x0ev\xdft\xb7[\xecp\xa8>\x8dd7\xe2\xd7\xc2<YI$.'\x82\x01X OT\x07\x07\\\xe5\xdf\x9bb\x08\x95o\xf1W#\xe1k9Y\x1env?[\xc2W68q\xe4\x07O"`

cbor2.loads(): `b"\xa2\x19\x01\x94\x83X \xf7\n\xb0*q\x15\x1d\x04\xedP\t\x90|\xc3\xb2\xcb,~\x9az\x8f\xb3\xf6j~\xe7<\xd5\xb7\xed+\xf8xHce:2.12:7753f830cdbaf25b07902e2b0fddd8c1fe196b04a8625629a39c738ebc7b7486X \x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x19\x01\x95\x82\x82\x01X \xf4^\x91e\xfa\xfd\xa6\n$U\x0ev\xdft\xb7[\xecp\xa8>\x8dd7\xe2\xd7\xc2<YI$.'\x82\x01X OT\x07\x07\\\xe5\xdf\x9bb\x08\x95o\xf1W#\xe1k9Y\x1env?[\xc2W68q\xe4\x07O"`

cbor2.loads(cbor2.loads()): `{404: [b'\xf7\n\xb0*q\x15\x1d\x04\xedP\t\x90|\xc3\xb2\xcb,~\x9az\x8f\xb3\xf6j~\xe7<\xd5\xb7\xed+\xf8', 'ce:2.12:7753f830cdbaf25b07902e2b0fddd8c1fe196b04a8625629a39c738ebc7b7486', b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'], 405: [[1, b"\xf4^\x91e\xfa\xfd\xa6\n$U\x0ev\xdft\xb7[\xecp\xa8>\x8dd7\xe2\xd7\xc2<YI$.'"], [1, b'OT\x07\x07\\\xe5\xdf\x9bb\x08\x95o\xf1W#\xe1k9Y\x1env?[\xc2W68q\xe4\x07O']]}`

File: src/node/historical_queries_adapter.cpp (hunk 6 above)
Comment (maxtropets) [reply src/node/historical_queries_adapter.cpp]: Talked offline, we can create a vector in place instead of a stack buffer

Comment (maxtropets) [reply src/node/historical_queries_adapter.cpp]: Done: https://github.com/microsoft/CCF/pull/6480/commits/0a2105cfc86f8c904b1c1c34a661ae0a43b455bd

File: src/crypto/openssl/cose_sign.cpp [hunk 9]
```diff
@@ -216,8 +217,16 @@ namespace ccf::crypto
         fmt::format("Can't finish QCBOR encoding with error code {}", err));
     }
 
-    return {
-      static_cast<const uint8_t*>(signed_cose.ptr),
-      static_cast<const uint8_t*>(signed_cose.ptr) + signed_cose.len};
+    if (signed_cose.ptr != underlaying_buffer.data())
+    {
+      LOG_FAIL_FMT("COSE sign1: QCBOR buffer doesn't match underlaing buffer");
```
Comment (achamayou) [inline src/crypto/openssl/cose_sign.cpp:222]:
```suggestion
      LOG_FAIL_FMT("COSE sign1: QCBOR buffer doesn't match underlying buffer");
```

Comment (achamayou) [reply src/crypto/openssl/cose_sign.cpp:222]: That does not seem like something we ought to check at runtime, it's either a guarantee the library offers (and which we can confirm in unit tests if necessary), or it's not and we should not rely on it.

File: src/crypto/openssl/cose_sign.cpp [hunk 10]
```diff
@@ -216,8 +217,16 @@ namespace ccf::crypto
...
+      return std::vector<uint8_t>{
+        static_cast<const uint8_t*>(signed_cose.ptr),
+        static_cast<const uint8_t*>(signed_cose.ptr) + signed_cose.len};
+    }
+
+    underlaying_buffer.resize(signed_cose.len);
+    underlaying_buffer.shrink_to_fit();
+    return underlaying_buffer;
```
Comment (achamayou) [inline src/crypto/openssl/cose_sign.cpp:230]:
```suggestion
    return underlying_buffer;
```

**Review by achamayou — APPROVED** [review]

**Pull request merged**